### PR TITLE
Added extra GC call before analyzing subgraph during freeze operation…

### PR DESCRIPTION
**Pull request by LepilkinaElena**
…s to decrease analyzed objects number

### CODE REVIEW

File: runtime/src/main/cpp/Memory.cpp [hunk 1]
```diff
@@ -2394,6 +2404,9 @@ bool clearSubgraphReferences(ObjHeader* root, bool checked) {
   auto state = memoryState;
   auto* container = root->container();
 
+  // Free cyclic garbage to decrease number of analyzed objects.
+  checkIfForceCyclicGcNeeded(state);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp:2408]: It makes sense to place this after  `if (isShareable(container))`.

File: runtime/src/main/cpp/Memory.cpp [hunk 2]
```diff
@@ -1997,6 +1997,16 @@ inline void checkIfGcNeeded(MemoryState* state) {
   }
 }
 
+inline void checkIfForceCyclicGcNeeded(MemoryState* state) {
+  if (state != nullptr && state->toFree->size() > kMaxToFreeSizeThreshold) {
+    // To avoid GC trashing check that at least 10ms passed since last GC.
+    if (konan::getTimeMicros() - state->lastGcTimestamp > 10 * 1000) {
+      GC_LOG("Calling GC from checkIfForceCyclicGcNeeded: %d\n", state->toFree->size())
+      garbageCollect(state, true);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp:2005]: I guess forced cycle collection shouldn't trigger increasing collect cycles threshold in `garbageCollect`.

Comment (LepilkinaElena) [reply runtime/src/main/cpp/Memory.cpp:2005]: Only collect cycles threshold ?
I now think about same logic for gcThreshold.

Comment (SvyatoslavScherbina) [reply runtime/src/main/cpp/Memory.cpp:2005]: Well, makes sense.

Comment (LepilkinaElena) [reply runtime/src/main/cpp/Memory.cpp:2005]: I made such logic only for increasing based on time.

File: runtime/src/main/cpp/Memory.cpp [hunk 3]
```diff
@@ -1997,6 +1997,16 @@ inline void checkIfGcNeeded(MemoryState* state) {
   }
 }
 
+inline void checkIfForceCyclicGcNeeded(MemoryState* state) {
+  if (state != nullptr && state->toFree->size() > kMaxToFreeSizeThreshold) {
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp:2001]: It might make sense to check `toFree != nullptr` too.

**Pull request merged**